levels/level_manager.go

```diff
@@ -1707,6 +1727,84 @@ func containsTable(seg *segment, tabID sst.SSTableID) bool {
 	return false
 }
 
+func getSegmentForRegistration(segmentEntries []segmentEntry, registration RegistrationEntry) int {
+	n := len(segmentEntries)
+	// if n == 0 then sort.Search considers 0 a legal insertion point
+	// for now we still return -1 to fit in with the current code
+	if n == 0 {
+		return -1
+	}
+	// fits the case i == 0 && i == len(segmentEntries)-1
+	if n == 1 {
+		return 0
+	}
+	// with sort.Search we need a compare function that partitions the array into true/false
+	index := sort.Search(n, func(i int) bool {
+		return i == n-1 || bytes.Compare(registration.KeyEnd, segmentEntries[i+1].rangeStart) < 0
+	})
+	// now we can check if the KeyStart is > than segmentEntries[i-1].rangeEnd
+	if index == 0 {
+		return index
+	}
+	if bytes.Compare(registration.KeyStart, segmentEntries[index-1].rangeEnd) > 0 {
+		if index == n {
+			return index - 1
+		} else {
+			return index
+		}
+	}
+	return -1
+}
+
+func getSegmentEntryForDeregistration(segmentEntries []segmentEntry, deRegistration RegistrationEntry) int {
+	n := len(segmentEntries)
+	if n == 0 {
+		return -1
+	}
+	index := sort.Search(n, func(i int) bool {
+		return bytes.Compare(deRegistration.KeyStart, segmentEntries[i].rangeStart) < 0 ||
+			(bytes.Compare(deRegistration.KeyStart, segmentEntries[i].rangeStart) >= 0 &&
+				bytes.Compare(deRegistration.KeyEnd, segmentEntries[i].rangeEnd) <= 0)
+	})
+	// if index < n it's possible that sort.Search will consider it a legal index
```

Comment:
I thought that sort.Search will return n if not found and something else if found? So not sure why this extra check is necessary. I could be missing something.

Reply:
`sort.Search` only finds the insertion point, so the extra check is required to make sure it actually found the element that's being searched for. The comment I left there is definitely not very clear on that point, now that I read it again.